### DISCUSSION TITLE
F2b misses a timestamp that exists

### DISCUSSION THREAD

**Discussion by jim-moe** [fail2ban/fail2ban#3783] [General]
fail2ban v1.1.0.1
opensuse LEAP 15.5
linux v5.14.21-150500.55.65-default x86_64

A filter has two regexes; the first matches as expected, the second does not. fail2ban-regex has no problem matching the same log entries.
Why is the second regex failing in the server?

The error message in the f2b log.
```
2024-07-01 12:24:27,142 fail2ban.filter [9089]: WARNING [named-1] Match without a timestamp: 01-Jul-2024 12:24:27.130 resolver: notice: DNS format error from 8.8.4.4#53 resolving cs.ffbtas.com/AAAA for 192.168.69.113#54170: Name . (SOA) not subdomain of zone com -- invalid response
```

The filter. The commented text is samples of what it is detecting.
```
# 18-Dec-2023 10:11:29.025 resolver: notice: DNS format error from 144.202.95.40#53 resolving phc.com/TXT for 127.0.0.1#46943: reply has no answer
# 20-Dec-2023 09:57:49.548 resolver: notice: DNS format error from 138.197.103.178#53 resolving net-up.com/TXT for 127.0.0.1#59039: Name . (SOA) not subdomain of zone net-up.com -- invalid response

failregex = ^.*\..* resolver\: notice\: DNS format error from <HOST>\#.* reply has no answer.*$
            ^.*\..* resolver\: notice\: DNS format error from <HOST>\#.* \-\- invalid response.*$

datepattern = %%d/%%b/%%Y %%H:%%M:%%S
```

**Comment by sebres** [comment]
> `18-Dec-2023 10:11:29.025 resolver: ...`
> `datepattern = %%d/%%b/%%Y %%H:%%M:%%S`

The timestamp in log can't be found with specified `datepattern` (uses `/` instead of `-` as separator). 
The correct one would be something like this:
```diff
- datepattern = %%d/%%b/%%Y %%H:%%M:%%S
+ datepattern = ^%%d-%%b-%%Y %%H:%%M:%%S(?:\.%%f)?
```
(`^` - just anchoring it to begin of line, and `(?:\.%%f)?` optionally matches milliseconds fraction).

Also note, that one of the default `datepattern`s should match too, so one doesn't really need to specify it (excepting for performance reasons):
https://github.com/fail2ban/fail2ban/blob/59c5e78ce90a49a6f5fb325d35199923c9c65f7c/fail2ban/server/datedetector.py#L149

By the way, your REs looks a bit vulnerable to me (e. g. catch-all breaking anchoring and that from both sides), something like that would be better (and as single RE, so additionally faster):
```ini
failregex = ^\S*\s+resolver: notice: DNS format error from <ADDR>#\d+ resolving \S* for \S+:.*(?:reply has no answer|-- invalid response)\s*$

datepattern = %%d/%%b/%%Y %%H:%%M:%%S
```

**Comment by jim-moe** [comment]
> The timestamp in log can't be found with specified datepattern (uses / instead of - as separator).
> The correct one would be something like this:
> 
>   -datepattern = %%d/%%b/%%Y %%H:%%M:%%S
>  +datepattern = ^%%d-%%b-%%Y %%H:%%M:%%S(?:\.%%f)?
> 

Ah, how obvious. Thanks for the tip.
